### PR TITLE
Add feature flag for merged cacerts secret

### PR DESCRIPTION
**Pull request by jaellio**
**Please provide a description of this PR:**

Makes using `cacerts` Secret for istiod generated root certs opt-in. All functionality related to #41507 will
now be opt-in. Considering making this default behavior for ambient.

Related comment: https://github.com/istio/istio/pull/46729#issuecomment-1740027815

### CODE REVIEW

**Comment by jaellio** [comment]
Alternative feature flags:

1. Convert existing `istio-ca-secret` to `cacerts`: `CONVERT_ISTIO_CA_SECRET_IF_EXISTS` (or something like that)
    - TRUE - if `istio-ca-secret` exists and there is no `cacerts` secret istiod will create a cacerts secret and populate it with the data from `istio-ca-secret`. **This is not yet implemented** but will be the riskiest functionality of the secret merging work and would benefit from being behind a feature flag. Alternatively, the user could create the `cacerts` manually from `istio-ca-secret`. 
    - FALSE - if a `istio-ca-secret exists and there is no `cacerts` secret does nothing.
  
    **Benefit** - better user experience. User doesn't have to enable a flag to create a `cacerts` secret on a fresh install. Hides the riskiest functionality.
    **Downside** - less safe than the all-inclusive option. Exposes a half solution to users.

2. (CURRENT APPROACH) Hide all `cacerts` merging behind a feature flag: `USE_CACERTS`
    - TRUE - (given the currently merged changes) If an `istio-ca-secret` or `cacerts` secret does not exist create a `cacerts` secret for the istio-generated self-signed secret. Eventually, the flag would expand to include istio merging the `istio-ca-secret` and creating a `cacerts` secret as described in option 1.
    - FALSE - follow the existing default workflow.
  
    **Benefit** - safe for users
    **Downside** - worse user experience. Potentially poor user adoption

File: security/pkg/pki/ca/ca.go [hunk 1]
```diff
@@ -119,9 +119,18 @@ type IstioCAOptions struct {
 // NewSelfSignedIstioCAOptions returns a new IstioCAOptions instance using self-signed certificate.
 func NewSelfSignedIstioCAOptions(ctx context.Context,
 	rootCertGracePeriodPercentile int, caCertTTL, rootCertCheckInverval, defaultCertTTL,
-	maxCertTTL time.Duration, org string, dualUse bool, namespace string, client corev1.CoreV1Interface,
+	maxCertTTL time.Duration, org string, useCacertsSecretName, dualUse bool, namespace string, client corev1.CoreV1Interface,
 	rootCertFile string, enableJitter bool, caRSAKeySize int,
 ) (caOpts *IstioCAOptions, err error) {
+	// TODO(jaellio): Implement istio-ca-secret to cacerts Secret migration. If only a istio-ca-secret
+	// exists, create a new cacerts secret and load the cacerts data. Currently, if
+	// USE_CACERTS_FOR_SELF_SIGNED_CA is enabled we ignore an existing istio-ca-secret and create a
```
Comment (hzxuzhonghu) [inline security/pkg/pki/ca/ca.go:127]: We need to create both two secrets named `istio-ca-secret` and `cacerts` to support forward compatibility.  Let's think about whether it is a must

**Comment by hzxuzhonghu** [comment]
IIUC @costinm means we should make sure to upgrade from 1.19  to 1.20 and downgrade from 1.20 to 1.19 works.

I am almost sure both these scenarios are already well supported. 

For 1.20, self signed certs is named `cacerts` as well, so if downgrade to 1.19, it can be mounted to 1.19 istiod, and istiod will use it seamlessly. 

Upgrade from 1.19 to 1.20 is natively supported.

I think the tricky point is when/how to smoothly get rid of `istio-ca-secret`.

**Comment by jaellio** [comment]
@hzxuzhonghu 
> IIUC @costinm means we should make sure to upgrade from 1.19 to 1.20 and downgrade from 1.20 to 1.19 works.
> 
> I am almost sure both these scenarios are already well supported.
> 
> For 1.20, self signed certs is named `cacerts` as well, so if downgrade to 1.19, it can be mounted to 1.19 istiod, and istiod will use it seamlessly.

The tricky downgrade scenario (with the secret merging work that has already been merged) is documented here https://docs.google.com/document/d/1jj7ZWhjgeQRNrCyaG6ah4A6XgYGc_r4bhIPMylSN9Qo/edit.
> 
> Upgrade from 1.19 to 1.20 is natively supported.
> 
> I think the tricky point is when/how to smoothly get rid of `istio-ca-secret`.

Do you want to get rid of `istio-ca-secret`? I don't think we want to delete any secret. Rather we just want to create a `cacerts` secret using that data, use an existing `cacerts`, or generate a new cert and store it in `cacerts`.

File: security/pkg/pki/ca/ca.go (hunk 1 above)
Comment (jaellio) [reply security/pkg/pki/ca/ca.go:127]: Not sure I understand. Are you saying creating both secrets is a must? What do we want this feature flag to mean/ protect?

**Comment by hzxuzhonghu** [comment]
>Do you want to get rid of istio-ca-secret? I don't think we want to delete any secret. Rather we just want to create a cacerts secret using that data, use an existing cacerts, or generate a new cert and store it in cacerts.

Yeah, otherwise, I do not see the meaning of what we do. we can copy that previously generated from `istio-ca-secret` to `cacerts`.

File: pilot/pkg/bootstrap/istio_ca.go [hunk 2]
```diff
@@ -491,7 +491,7 @@ func (s *Server) createSelfSignedCACertificateOptions(fileBundle *ca.SigningCAFi
 		caOpts, err = ca.NewSelfSignedIstioCAOptions(ctx,
 			selfSignedRootCertGracePeriodPercentile.Get(), SelfSignedCACertTTL.Get(),
 			selfSignedRootCertCheckInterval.Get(), workloadCertTTL.Get(),
-			maxWorkloadCertTTL.Get(), opts.TrustDomain, true,
+			maxWorkloadCertTTL.Get(), opts.TrustDomain, features.UseCacertsForSelfSignedCA, true,
```
Comment (costinm) [inline pilot/pkg/bootstrap/istio_ca.go:494]: as a general comment (for future PRs maybe): when a function has 10 params - it's good time to define a struct...

**Review by costinm — APPROVED** [review]
It has a feature flag, off by default - looks safe to me.

I'm not so concerned about 1.19 - 1.20 - but skip version ( which are pretty common ), but that should be fine as long as the skip-version is not combined with enabling the flag. 

Unless I missed something - looks good, and we can do the cleanup and enable by default in a couple of releases.

**Pull request merged**